### ISSUE TITLE
Does YOLOv5 have the ability to detect crowd objects?

### ISSUE THREAD

**Issue by Nise-2-meet-U** [ultralytics/yolov5#1779]
## ❔Question
I know there are methods to count crowd objects, like MCNN..

But I want to know what is the limit of yolov5. like train a big model e.g. yolov5x, will it work?

pics like below..
![IMG_56](https://user-images.githubusercontent.com/75665054/103137059-9f1d4600-4700-11eb-91c7-4a10c5bc673f.jpg)
![IMG_20](https://user-images.githubusercontent.com/75665054/103137062-a5132700-4700-11eb-9fb6-9ba56a846d96.jpg)
![IMG_16](https://user-images.githubusercontent.com/75665054/103137063-a8a6ae00-4700-11eb-8d60-be7ebcf3c353.jpg)

**Comment by glenn-jocher** [comment]
@Nise-2-meet-U depends on your definition of 'work'. 

All solutions will work to some degree, whether YOLOv5 is the best solution is not knowable without empirical results obtained from experimenting with the various solutions, though I do know that there are more specific crowd-counting solutions that are more tailer-made for this task that approach the problem as a heatmap rather than a detection task.

I think if you wanted to customize a detection architecture to this, your training data is paramount. Based on your images, you want a 'head' class, and then at inference time you set your NMS threshold very high, i.e. 0.90. If you take these two steps I don't see why you shouldn't be able to produce excellent results here.

**Comment by Nise-2-meet-U** [comment]
> @Nise-2-meet-U depends on your definition of 'work'.
> 
> All solutions will work to some degree, whether YOLOv5 is the best solution is not knowable without empirical results obtained from experimenting with the various solutions, though I do know that there are more specific crowd-counting solutions that are more tailer-made for this task that approach the problem as a heatmap rather than a detection task.
> 
> I think if you wanted to customize a detection architecture to this, your training data is paramount. Based on your images, you want a 'head' class, and then at inference time you set your NMS threshold very high, i.e. 0.90. If you take these two steps I don't see why you shouldn't be able to produce excellent results here.

Thanks for your reply.
In the crowd "head" scenario, set NMS threshold to 0.90 could get overlaped boxes like this.
![111](https://user-images.githubusercontent.com/75665054/103145938-b6017e00-477c-11eb-8123-4567a7b368c2.png)
I have trained on my dataset using yolov5s parameters width_multiple=0.5, depth_multiple=0.33. 
But results are not so good, my expriment-images are listed below:
![微信截图_20201226132325](https://user-images.githubusercontent.com/75665054/103145989-90c13f80-477d-11eb-881f-8c0d4082eaeb.png)
![微信截图_20201226132815](https://user-images.githubusercontent.com/75665054/103146037-4c826f00-477e-11eb-8348-e53a75d4a5ad.png)
![微信截图_20201226132831](https://user-images.githubusercontent.com/75665054/103146039-4f7d5f80-477e-11eb-84da-9cdac83c443a.png)
Really crowed regin have no detected head.

It seems one cannot detect really small objects(smaller than 15*15 pixels). I even tried train a larger model(width_multiple=1, depth_multiple=1), sadly, results are similar.

**Comment by glenn-jocher** [comment]
@Nise-2-meet-U looks pretty good to me, but I'm not a domain expert in crowd detection either. A few points to consider:

- results will be a function of your training data, therefore abstract performance on your anecdotal images is difficult to quantify independently the context of your training data. In laymans terms detecting far away people requires a dataset labelled as such, or more rigorous scale jitter with accompanying label threshold modifications in the candidate box selection.
- 0.90 is simply a starting point, it is up to the user to tune a satisfactory inference parameter space.
- One trick you might use is to perform NMS on reduced size boxes, i.e. divide width and height by 2 on the NMS inputs. This should allow you to reduce iou threshold to produce better results.
- This type of domain may call for a completely different NMS alternative, as your priorities are not bounding boxes but simply counting objects. 
- There is a maximum object count threshold in the NMS function. Default is 300. You may want to revisit the default.

**Comment by Nise-2-meet-U** [comment]
> @Nise-2-meet-U looks pretty good to me, but I'm not a domain expert in crowd detection either. A few points to consider:
> 
> * results will be a function of your training data, therefore abstract performance on your anecdotal images is difficult to quantify independently the context of your training data. In laymans terms detecting far away people requires a dataset labelled as such, or more rigorous scale jitter with accompanying label threshold modifications in the candidate box selection.
> * 0.90 is simply a starting point, it is up to the user to tune a satisfactory inference parameter space.
> * One trick you might use is to perform NMS on reduced size boxes, i.e. divide width and height by 2 on the NMS inputs. This should allow you to reduce iou threshold to produce better results.
> * This type of domain may call for a completely different NMS alternative, as your priorities are not bounding boxes but simply counting objects.
> * There is a maximum object count threshold in the NMS function. Default is 300. You may want to revisit the default.

I have changed the default NMS setting from 300 to 30000.
The trainning labels are correct as far as I can tell.
Thank you, anyway. You have done an amazing job! Yolov5 is the best detection so far!

**Comment by glenn-jocher** [comment]
@Nise-2-meet-U no problem. You may want to increase your scale hyperparameter also to force the dataloader to scale to more extremes, i.e. from 0.5 to 0.9, which will scale objects to 0.1x - 1.9x their original size during training.

**Comment by GianniGi** [comment]
Are there pretrained models that can count very tight crowds of people? Does yolo work or you have to resort to MCNN?

**Comment by glenn-jocher** [comment]
@GianniGi there's a wide array of options, i.e. https://github.com/deepakcrk/yolov5-crowdhuman

**Comment by GianniGi** [comment]
@glenn-jocher 
> @GianniGi there's a wide array of options, i.e. https://github.com/deepakcrk/yolov5-crowdhuman

thanks, I see that there are a lot of repositories, none seems updated or predominant upon the others though

**Comment by GianniGi** [comment]
> map-based model

what repository would you sugges, in this case? thanks

**Comment by glenn-jocher** [comment]
@GianniGi You can check out the CSRNet repository at https://github.com/leeyeehoo/CSRNet-keras. It's worth experimenting with different architectures and methods to find what works best for your specific use case. Good luck!
